feat(search): component library search component

dewen opened the pull request:
<!--
  Before issuing a pull request:

  - Please read our contribution guidelines (https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-documentation/doc/Development/Contributing.md), especially the section on Pull Requests.
  - Please first create an issue (https://github.com/johnsonandjohnson/Bodiless-JS/issues/new). All pull requests must be linked to an issue.
-->

<!--
  IMPORTANT
  
  The pull request title will become the commit message title at merge, and
  should adhere to Angular Commit Message Conventions. Please see (https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-documentation/doc/Development/Contributing.md) for details and examples.
-->

## Changes
<!-- Brief description of the changes introduced by this PR -->

- Create wrapper class on Lunr in-browser search library
- Create SearchTool for generating search index.
- Search related components for site search feature.

## Test Instructions
<!-- Detailed description of how this feature was (and should be) tested
  - Setup instructions (if any)?
  - List of test cases, with steps and expected results?
  - List
-->

- Edit search configures in site .env.site file
- Deploy test-site and review the search function.
- Add test content to pages, rebuild static site. Search test content.
- Check AC 1-4 of original issue.

## Related Issues
<!--
  Link to the issue that is fixed or resolved by this PR (if there is one)
  e.g. Fixes #1234, Closes #4567

  Link to an issue that is partially addressed by this PR (if there are any)
-->
fixes #204 

<!-- IF THIS PR INTRODUCES A BREAKING CHANGE

BREAKING CHANGE: Describe the nature of the breaking change here.

More Details about the breaking change.
-->

beliayeu commented on examples/test-site/package.json:
```diff
@@ -76,7 +78,8 @@
     "psh-update": "node ./lib/scripts/bodiless-psh-update",
-    "psh-validate": "node ./lib/scripts/bodiless-psh-validate"
+    "psh-validate": "node ./lib/scripts/bodiless-psh-validate",
+    "search-index": "node ./node_modules/@bodiless/search/lib/scripts/bodiless-search-index.js"
```
suggest making bodiless-search-index as bin so that we can execute without specifying relative path

beliayeu replied:
probably just having build:search is sufficient

beliayeu commented on examples/test-site/src/components/Search/index.tsx:
```diff
@@ -0,0 +1,49 @@
+
+const Icon = flow(
+  addClasses('material-icons cursor-pointer align-middle'),
+)(stylable((props: any) => (<i {...props}>{props.children}</i>)));
```
I believe we can just use I component imported from bodiless/fclasses

beliayeu commented on examples/test-site/src/components/Search/index.tsx:
```diff
@@ -0,0 +1,49 @@
+    <Icon>{icon}</Icon>
+  </Component>
+);
+const withSearchButton = (icon: string) => flow(
```
tentatively

```
withSearchButtonIcon = withIcon('Search')
```

search design will be
```
const searchDesign = {
  SearchBox: addClasses('px-2 align-middle text-1xl'),
  SearchButton: withSearchButtonIcon,
};
```

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,119 @@
+    const { sourcePath, sourceTypes } = settings;
+    // @todo: glob file list
+
+    const path$ = path.resolve(process.cwd(), sourcePath);
```
discuss allowing site builders to override sourcePath

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,119 @@
+    const html = fs.readFileSync(path.resolve(sourcePath, filePath)).toString();
+    const $ = cheerio.load(html);
+    return {
+      id: v1(),
```
discuss allowing site builders to define own rules

dewen replied:
yes, added some selector rule and path option.

beliayeu commented on packages/bodiless-search/src/types.ts:
```diff
@@ -0,0 +1,65 @@
+ * limitations under the License.
+ */
+
+import lunr from 'lunr';
```
does lunr provides types that we can leverage?

dewen replied:
Yes, I was including some types from @types/lunr, i.e. lunr.Index from types.ts

beliayeu commented on examples/test-site/src/components/Search/index.tsx:
```diff
@@ -0,0 +1,66 @@
+
+const Icon = flow(
+  addClasses('material-icons cursor-pointer align-middle'),
+)(stylable((props: any) => (<i {...props}>{props.children}</i>)));
```
recommend using I component from fclasses

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,144 @@
+import mime from 'mime';
+import path from 'path';
+import { v1 } from 'uuid';
+import {
```
I believe it is better to use `import type` for type importing

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,144 @@
+    this.searchEngine.addDocuments(documents);
+
+    const ind = this.searchEngine.exportIndex();
+    const targetPath$ = path.resolve(process.cwd(), targetPath);
```
do we want to couple to process.cwd?

dewen replied:
This is backend path for read/write the source content, using absolute path.

Any concerns?

dewen replied:
It's user's option to choose between absolute or relative path. So in case of relative path, the cwd is needed.

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,144 @@
+
+    const path$ = path.resolve(process.cwd(), sourcePath);
+    if (!fs.existsSync(path$)) {
+      throw new Error('Invalid source path');
```
nice to have - add path$ to the error page - 
```
throw new Error(`Invalid source path ${path$}`);
```

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,144 @@
+      throw new Error('Invalid source path');
+    }
+
+    const pattern = `**/+(${sourceTypes.map(v => `*.${v}`).join('|')})`;
```
discuss allowing to configure the whole pattern on site level (can be useful to exclude pages from indexing)

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,144 @@
+        switch (mimeType) {
+          case 'text/html': {
+            const html = fs.readFileSync(path.resolve(sourcePath, filePath)).toString();
+            const doc = this.htmlToDocument(html, selector, exclude);
```
discuss allowing configuring it on site level (allow site builders to define their own rules to extract text from html)

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,258 @@
+  SearchButton: ComponentType<any>;
+};
+
+type SearchBoxProps = {
```
do we need to have a separate type? per my understanding HTMLProps<HTMLInputElement>> covers it

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,258 @@
+const searchClient = new SearchClient();
+
+const SearchBoxBase: FC<SearchBoxProps & HTMLProps<HTMLInputElement>> = props => (
+  <Input {...props} placeholder="Search" />
```
discuss allowing placeholder override on site level

hvanyo replied:
@dewen  @beliayeu  -- agree -- we need to allow this be translatable and exposed to site builder to override.

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,258 @@
+    if (
+      searchPagePath !== window.location.pathname.replace(/^\//, '').replace(/\/$/, '')
+    ) {
+      window.location.href = `/search?q=${queryString}`;
```
take /search from searchPagePath?

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,258 @@
+
+  const onClickHandler = (event: React.MouseEvent) => {
+    event.preventDefault();
+    if (searchPagePath !== window.location.pathname.replace(/^\//, '').replace(/\/$/, '')) {
```
logic duplication with searchHandler

hvanyo commented on examples/test-site/src/components/Search/index.tsx:
```diff
@@ -0,0 +1,66 @@
+);
+
+const searchDesign = {
+  SearchWrapper: addClasses('my-4 border bl-border-black align-middle'),
```
@dewen  for all these classes, can you make them in tokens and then put in token.ts in Search Folder.   Take a look at https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/examples/test-site/src/components/ProductTout. and within index.tsx is more funtion and all the styling are tokens in token.ts.   Also try use tokens if possible from elements.token.ts (if they work) but you have some that you need to create here as well.

hvanyo commented on examples/test-site/src/components/Search/index.tsx:
```diff
@@ -0,0 +1,66 @@
+  SearchResultList: addClasses('py-2'),
+  SearchResultSummary: addClasses('text-sm italic'),
+  SearchResultListItem: withDesign({
+    ItemAnchor: addClasses('my-4 bl-text-blue-500 underline'),
```
Any classes within test site that is defining look of component on site should not be prefaced with 'bl-'  The exception here is component is adding editor UX form/window.

hvanyo commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,258 @@
+  const {
+    SearchResultWrapper, SearchResultList, SearchResultListItem, SearchResultSummary,
+  } = components;
+  const showResultCount = `Showing ${searchResultContext.results.length} results.`;
```
@dewen  another translated item that needs to be exposed to site builder.

hvanyo commented:
@beliayeu  @dewen  this is not technically in acs :( but while you are working on this think about imodium case we have as two languages in one.  
1) need translatable strings on exposed strings
2) All language content indexed together (and possibly sorted automatically when they search for english or arabic phrase) -- i.e. if you search for "stomach" it naturally only returns english results with stomach or if you search for " الإسهال" it only returns arabic results.
3) two separate indexes for each language -- doubt it..  

Anyway, its not AC requirements but please consider it in design/build phase and or what we need to do support more languages -> https://lunrjs.com/guides/language_support.html

dewen commented on examples/test-site/.env.site:
```diff
@@ -52,3 +52,10 @@ SITE_URL='https://www.example.com/'
 # To enable it - uncomment the following line.
 # BODILESS_ALERT_ON_PAGE_LOAD_ENABLED='1'
+
+BODILESS_SEARCH_PAGE='search'
```
I think search page configure should come from one place. As env variable maybe the best place, but since site builder can add multiple search box on a page, i.e. one in header, another one in body, pass the search page path can be redundant?

beliayeu replied:
@dewen, makes sense, resolving conversation

dewen commented on examples/test-site/src/components/Search/index.tsx:
```diff
@@ -0,0 +1,66 @@
+  asSimpleSearchResult,
+)(SearchResultClean);
+
+export const InlineSearchBox = flow(asInlineSearch)(SearchClean);
```
It throws TS error if flow removed:

```
Argument of type 'ComponentType<SearchProps>' is not assignable to parameter of type 'ComponentType<DesignableProps<DesignableComponents>>'.
  Type 'ComponentClass<SearchProps, any>' is not assignable to type 'ComponentType<DesignableProps<DesignableComponents>>'.
    Type 'ComponentClass<SearchProps, any>' is not assignable to type 'ComponentClass<DesignableProps<DesignableComponents>, any>'.
      Types of property 'propTypes' are incompatible.
        Type 'WeakValidationMap<SearchProps>' has no properties in common with type 'WeakValidationMap<DesignableProps<DesignableComponents>>'.
```

malachye commented on packages/bodiless-search/README.md:
```diff
@@ -0,0 +1,167 @@
+# Search Component
+
+Bodiless Search component provides in-borswer full-text search for Bodiless site build. By default, it uses [Lunr](https://lunrjs.com/) search library for creating index and performing the search behind the sence. Site builders could also switch to their own customized search library during site build.
```
```suggestion
The Bodiless Search component provides in-browser, full-text search for Bodiless sites. By default, it uses [Lunr](https://lunrjs.com/) search library for creating index and performing the search behind the scenes. Site builders can also switch to their own customized search library during site build.
```

malachye commented on packages/bodiless-search/README.md:
```diff
@@ -0,0 +1,167 @@
+
+## Usage
+
+To enable the search on a site project, site builder needs to add search configures inside environment setting file. Then, creates design HOC functions that add styles to search components. Insert the sytled search box and search result components into the pages for renderring.
```
```suggestion
To enable the search on a site project, the site builder needs to add search configurations inside the environment setting file. Next create the design HOC functions, which will add styles to search components. Insert the styled search box and search result components into the pages for rendering.
```

malachye commented on packages/bodiless-search/README.md:
```diff
@@ -0,0 +1,167 @@
+
+### Configuration
+
+Set the following configure options in .env.site base on search requirement.
```
```suggestion
Set the following configuration options in `.env.site` based on search requirement.
```

malachye commented on packages/bodiless-search/README.md:
```diff
@@ -0,0 +1,167 @@
+
+### Create Search Index
+
+Then add following line to `script` section of package.json. 
```
```suggestion
Add the following line to `script` section of package.json. 
```

malachye commented on packages/bodiless-search/README.md:
```diff
@@ -0,0 +1,167 @@
+
+The "Search*Clean" alias given to imported components is a Bodiless convention for unstyled package components. Site builder will need to format them before adding to pages.
+
+For how to add design to clean components, please refer documentation of [Bodiless Design System](https://johnsonandjohnson.github.io/Bodiless-JS/#/Design/DesignSystem).
```
```suggestion
For how to add design to clean components, please refer to the [Bodiless Design System](https://johnsonandjohnson.github.io/Bodiless-JS/#/Design/DesignSystem) documentation.
```

malachye commented on packages/bodiless-search/README.md:
````diff
@@ -0,0 +1,167 @@
+
+```
+
+To add `SimpleSearchResult` component, site builder will also need to import `withSearchResult` HOC, which can add `SearchResultProvider` context provider to a container component on the page, i.e.
````
```suggestion
To add `SimpleSearchResult` component, import `withSearchResult` HOC, which can add `SearchResultProvider` context provider to a container component on the page, i.e.
```

malachye commented on packages/bodiless-search/README.md:
````diff
@@ -0,0 +1,167 @@
+
+```
+
+Here, `SimpleSearchBox` and `SimpleSearchResult` are exported as site level components for search renderring. Site builder will need to add them on pages that require search feature. For example, to display the search box on page header, `SimpleSearchBox` can be added to `src/components/Layout/header.tsx`,
````
```suggestion
Here, `SimpleSearchBox` and `SimpleSearchResult` are exported as site level components for search rendering. They will need to be added on pages that require the search feature. For example, to display the search box on page header, `SimpleSearchBox` can be added to `src/components/Layout/header.tsx`,
```

beliayeu commented on packages/bodiless-search/src/scripts/bodiless-search-index.ts:
```diff
@@ -0,0 +1,50 @@
+import type { TSearchConf, TSearchIndexSettings } from '../types';
+
+require('dotenv').config({
+  path: '.env.development',
```
why .env.development?

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,139 @@
+          case 'text/html': {
+            const html = fs.readFileSync(path.resolve(sourcePath, filePath)).toString();
+            const doc = this.htmlToDocument(html, selector, exclude);
+            const filePathClean = filePath.replace(/index.html$/i, '');
```
tentatively, put it inside if condition
```
if (!doc.title) {
  const filePathClean = filePath.replace(/index.html$/i, '');
  doc.title = filePathClean;
}
```

also, we may want to cover a case when there is an html file different to index.html (for instance, products.html)

beliayeu commented on packages/bodiless-search/src/SearchTool.ts:
```diff
@@ -0,0 +1,139 @@
+   */
+  htmlToDocument = (html: string, selector: string, exclude: string): TDocument => {
+    const $ = cheerio.load(html);
+    const title = $('h1').text().trim() || $('title').text().trim();
```
@dewen, am I right that site builders won't be able to override rules for extracting title?

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,270 @@
+import { useSearchResultContext } from './SearchContextProvider';
+import { TSearchResult } from '../types';
+
+const querystring = require('query-string');
```
why require? not import?

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,270 @@
+    <SearchWrapper>
+      <SearchInput
+        value={queryString}
+        onChange={onChangeHandler}
```
@dewen, consider using memoized callbacks to prevent unnecessary renders  (https://reactjs.org/docs/hooks-reference.html#usecallback)

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -0,0 +1,270 @@
+    return (Date.now() <= expires);
+  };
+
+  const loadIndex = async () => {
```
to discuss splitting of this component,
as an option we can extract some functionality from this base component to a hoc
for instance, there will a hoc responsible for loading index

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -178,15 +178,61 @@ const useDidMountEffect = (func: Function[]) => {
+        const rawIndex = localStorage.getItem('search:index') || '{}';
+        let index = JSON.parse(rawIndex);
+        if (!validateIndex(index)) {
+          const response = await axios.get('/lunr.idx');
```
per my understanding we will get issues if site builders change BODILESS_SEARCH_INDEX_PATH configuration

beliayeu replied:
do we want to somehow validate response?

dewen replied:
You are right, I update the path variable to get from env settings.

dewen replied:
I think failed response does goto catch block.

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -178,15 +178,61 @@ const useDidMountEffect = (func: Function[]) => {
+      } catch (error) {
+        console.log('Failed to load search index file.');
+      }
+      const { q } = querystring.parseUrl(window.location.search).query;
```
why are we performing search in a component responsible for index loading?

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -264,7 +275,9 @@ const SearchBoxBase: FC<SearchProps> = ({ components, ...props }) => {
 };
 
-export const SearchBox = designable(searchComponents)(SearchBoxBase) as ComponentType<SearchProps>;
+export const SearchBox = designable(searchComponents)(
```
tentatively, add flow for better readability

beliayeu commented on packages/bodiless-search/src/components/SearchContextProvider.tsx:
```diff
@@ -13,30 +13,63 @@
+  const didMountRef = useRef(false);
+  const searchTermRef = useRef('');
+  useEffect(() => {
+    if (!didMountRef.current) {
```
I believe we should also check if searchTermRef.current is a default value since there might be a use case when child updates the term before useEffect is executed

TatsianaMiraniuk commented:
Hi @dewen ,

as discussed, below please see the issues that need to be fixed

**issue_1:**
A search term that a user typed in a search field is expected to be shown in 2nd search filed in /search page.
![Screen Shot 2020-10-30 at 10 23 31 AM](https://user-images.githubusercontent.com/3224040/97716542-213f1480-1a9a-11eb-9226-6851a3d4c966.png)

TatsianaMiraniuk commented:
**issue_2:**
Clicking a search icon when nothing in a search field that's located in the header opens /search/ page with 0 results. But doing the same in 2nd search field in /search/ page returns all pages in the results.

Per conversation, if a user doesn't enter any search term and clicks a magnifier icon, then /search/ page should show all pages as the results

TatsianaMiraniuk commented:
issue_3:
'Search' placeholder text is missing a padding at the left

TatsianaMiraniuk commented:
@dewen 

**issue_4:**
there is an error in a console ''Uncaught (in promise) IPu3.M.QueryParseError {name: "QueryParseError", message: "expecting term or field, found nothing", start: 0, end: 1'. if a search term contains any of the following chars
^ : - ~
![Screen Shot 2020-10-30 at 4 58 43 PM](https://user-images.githubusercontent.com/3224040/97756432-384d2900-1ad1-11eb-832f-9d973dc94484.png)

TatsianaMiraniuk commented:
@dewen ,
can you please figure out why the following happens? 

**issue_5:**

**Steps to reproduce:**
1. open Home page
2. enter J&J in the search field
3. click a magnifier icon/ or press enter

**Actual Results:**
Search page opens with 0 results

4. enter J&J in any search field in Search page
5. click a magnifier icon/ or press enter

**Actual results:**
Home page is returned in the results (as expected)

dewen commented:
> Hi @dewen ,
> 
> as discussed, below please see the issues that need to be fixed
> 
> **issue_1:**
> A search term that a user typed in a search field is expected to be shown in 2nd search filed in /search page.
> ![Screen Shot 2020-10-30 at 10 23 31 AM](https://user-images.githubusercontent.com/3224040/97716542-213f1480-1a9a-11eb-9226-6851a3d4c966.png)

fixed

dewen commented:
> **issue_3:**
> 'Search' placeholder text is missing a padding at the left

fixed

dewen commented:
> **issue_2:**
> Clicking a search icon when nothing in a search field that's located in the header opens /search/ page with 0 results. But doing the same in 2nd search field in /search/ page returns all pages in the results.
> 
> Per conversation, if a user doesn't enter any search term and clicks a magnifier icon, then /search/ page should show all pages as the results

fixed

dewen commented:
> @dewen ,
> can you please figure out why the following happens?
> 
> **issue_5:**
> 
> **Steps to reproduce:**
> 
> 1. open Home page
> 2. enter J&J in the search field
> 3. click a magnifier icon/ or press enter
> 
> **Actual Results:**
> Search page opens with 0 results
> 
> 1. enter J&J in any search field in Search page
> 2. click a magnifier icon/ or press enter
> 
> **Actual results:**
> Home page is returned in the results (as expected)

This is due to "&" encoding issue in url. Fixed

dewen commented:
> @dewen
> 
> **issue_4:**
> there is an error in a console ''Uncaught (in promise) IPu3.M.QueryParseError {name: "QueryParseError", message: "expecting term or field, found nothing", start: 0, end: 1'. if a search term contains any of the following chars
> ^ message: "boost must be numeric"
> : message: "unrecognised field ''"
> -- message: "expecting term or field, found nothing"
> ~ message: "edit distance must be numeric" or 'expected either a field or a term, found EDIT_DISTANCE'
> ![Screen Shot 2020-10-30 at 4 58 43 PM](https://user-images.githubusercontent.com/3224040/97756432-384d2900-1ad1-11eb-832f-9d973dc94484.png)

Lunr wildcard char replaced.

TatsianaMiraniuk commented:
@dewen ,

please see **issue_6:**
the text in a Title of a collapsed by default accordion is not searchable.
E.g. /products/product-a/ page has several collapsed by default accordions.
If we search by 'Storage', or 'Warnings',  Search returns 0 results.

Please note, that Search works as expected for a Body of a collapsed by default accordion.
Search also works ok if a term is from a Title of an open/expanded by default accordion.

TatsianaMiraniuk commented:
> > Hi @dewen ,
> > as discussed, below please see the issues that need to be fixed
> > **issue_1:**
> > A search term that a user typed in a search field is expected to be shown in 2nd search filed in /search page.
> > ![Screen Shot 2020-10-30 at 10 23 31 AM](https://user-images.githubusercontent.com/3224040/97716542-213f1480-1a9a-11eb-9226-6851a3d4c966.png)
> 
> fixed

the fix is tested. works as expected.
Thanks!

TatsianaMiraniuk commented:
> > **issue_2:**
> > Clicking a search icon when nothing in a search field that's located in the header opens /search/ page with 0 results. But doing the same in 2nd search field in /search/ page returns all pages in the results.
> > Per conversation, if a user doesn't enter any search term and clicks a magnifier icon, then /search/ page should show all pages as the results
> 
> fixed

the fix is checked, works as expected.

Thanks!

TatsianaMiraniuk commented:
> > **issue_3:**
> > 'Search' placeholder text is missing a padding at the left
> 
> fixed

the fix is checked. Thanks!

TatsianaMiraniuk commented:
> > @dewen
> > **issue_4:**
> > there is an error in a console ''Uncaught (in promise) IPu3.M.QueryParseError {name: "QueryParseError", message: "expecting term or field, found nothing", start: 0, end: 1'. if a search term contains any of the following chars
> > ^ message: "boost must be numeric"
> > : message: "unrecognised field ''"
> > -- message: "expecting term or field, found nothing"
> > ~ message: "edit distance must be numeric" or 'expected either a field or a term, found EDIT_DISTANCE'
> > ![Screen Shot 2020-10-30 at 4 58 43 PM](https://user-images.githubusercontent.com/3224040/97756432-384d2900-1ad1-11eb-832f-9d973dc94484.png)
> 
> Lunr wildcard char replaced.

@dewen ,

the errors in a console iare still reproducible for the following chars:
^
-
~
Can you please take a look?

TatsianaMiraniuk commented:
> issue_5



> > @dewen ,
> > can you please figure out why the following happens?
> > **issue_5:**
> > **Steps to reproduce:**
> > 
> > 1. open Home page
> > 2. enter J&J in the search field
> > 3. click a magnifier icon/ or press enter
> > 
> > **Actual Results:**
> > Search page opens with 0 results
> > 
> > 1. enter J&J in any search field in Search page
> > 2. click a magnifier icon/ or press enter
> > 
> > **Actual results:**
> > Home page is returned in the results (as expected)
> 
> This is due to "&" encoding issue in url. Fixed

Checked that the bug is not reproduced. Thanks!

TatsianaMiraniuk commented:
Hi @dewen ,

please see **issue_7:**
the following chars are not displayed in a body of a listing page
`~ % * + - : “`

E.g. search for
21:9
![Screen Shot 2020-11-02 at 4 03 36 PM](https://user-images.githubusercontent.com/3224040/97930694-f96ddc00-1d39-11eb-89c7-9dbf1c7dbce8.png)
or
50%
![Screen Shot 2020-11-02 at 6 34 23 PM](https://user-images.githubusercontent.com/3224040/97930766-1d312200-1d3a-11eb-8efd-c271f879b243.png)

  


Thanks!

dewen commented:
> Hi @dewen ,
> 
> please see **issue_7:**
> the following chars are not displayed in a body of a listing page
> `~ % * + - : “`
> 
> E.g. search for
> 21:9
> ![Screen Shot 2020-11-02 at 4 03 36 PM](https://user-images.githubusercontent.com/3224040/97930694-f96ddc00-1d39-11eb-89c7-9dbf1c7dbce8.png)
> or
> 50%
> ![Screen Shot 2020-11-02 at 6 34 23 PM](https://user-images.githubusercontent.com/3224040/97930766-1d312200-1d3a-11eb-8efd-c271f879b243.png)
> 
> fixed
> 
> Thanks!

fixed

dewen commented:
> > > @dewen
> > > **issue_4:**
> > > there is an error in a console ''Uncaught (in promise) IPu3.M.QueryParseError {name: "QueryParseError", message: "expecting term or field, found nothing", start: 0, end: 1'. if a search term contains any of the following chars
> > > ^ message: "boost must be numeric"
> > > : message: "unrecognised field ''"
> > > -- message: "expecting term or field, found nothing"
> > > ~ message: "edit distance must be numeric" or 'expected either a field or a term, found EDIT_DISTANCE'
> > > ![Screen Shot 2020-10-30 at 4 58 43 PM](https://user-images.githubusercontent.com/3224040/97756432-384d2900-1ad1-11eb-832f-9d973dc94484.png)
> > 
> > 
> > Lunr wildcard char replaced.
> 
> @dewen ,
> 
> the errors in a console iare still reproducible for the following chars:
> `^ - ~`
> Can you please take a look?

Fixed

dewen commented:
> @dewen ,
> 
> please see **issue_6:**
> q parameter in URL isn't changed if search happens in Search page
> 
> **Steps to reproduce:**
> 
> 1. open Home page
> 2. in Search field type any term, e.g. 'term1'
> 3. press Enter => Search page loads with q=term1 in URL
> 4. in Search field type other term, e.g. 'term2'
> 5. press Enter
> 
> **Actual Results:**
> q parameter stays with the old value in URL (q=term1)
> 
> **Expected Results:**
> q parameter reflects the actual search term (q=term2)

fixed and replaced with #

TatsianaMiraniuk commented:
> > > > @dewen
> > > > **issue_4:**
> > > > there is an error in a console ''Uncaught (in promise) IPu3.M.QueryParseError {name: "QueryParseError", message: "expecting term or field, found nothing", start: 0, end: 1'. if a search term contains any of the following chars
> > > > ^ message: "boost must be numeric"
> > > > : message: "unrecognised field ''"
> > > > -- message: "expecting term or field, found nothing"
> > > > ~ message: "edit distance must be numeric" or 'expected either a field or a term, found EDIT_DISTANCE'
> > > > ![Screen Shot 2020-10-30 at 4 58 43 PM](https://user-images.githubusercontent.com/3224040/97756432-384d2900-1ad1-11eb-832f-9d973dc94484.png)
> > > 
> > > 
> > > Lunr wildcard char replaced.
> > 
> > 
> > @dewen ,
> > the errors in a console iare still reproducible for the following chars:
> > `^ - ~`
> > Can you please take a look?
> 
> Fixed

the fix is confirmed - no errors in a console. Thanks!

TatsianaMiraniuk commented:
> > @dewen ,
> > please see **issue_6:**
> > q parameter in URL isn't changed if search happens in Search page
> > **Steps to reproduce:**
> > 
> > 1. open Home page
> > 2. in Search field type any term, e.g. 'term1'
> > 3. press Enter => Search page loads with q=term1 in URL
> > 4. in Search field type other term, e.g. 'term2'
> > 5. press Enter
> > 
> > **Actual Results:**
> > q parameter stays with the old value in URL (q=term1)
> > **Expected Results:**
> > q parameter reflects the actual search term (q=term2)
> 
> fixed and replaced with #

the if is checked. works as expected. Thanks!

TatsianaMiraniuk commented:
> > Hi @dewen ,
> > please see **issue_7:**
> > the following chars are not displayed in a body of a listing page
> > `~ % * + - : “`
> > E.g. search for
> > 21:9
> > ![Screen Shot 2020-11-02 at 4 03 36 PM](https://user-images.githubusercontent.com/3224040/97930694-f96ddc00-1d39-11eb-89c7-9dbf1c7dbce8.png)
> > or
> > 50%
> > ![Screen Shot 2020-11-02 at 6 34 23 PM](https://user-images.githubusercontent.com/3224040/97930766-1d312200-1d3a-11eb-8efd-c271f879b243.png)
> > fixed
> > Thanks!
> 
> fixed

The fix is checked for ~!@#$%^&*()_+=-`{}|\][:"';<>?/., - all chars are visible. Thanks!

beliayeu commented on packages/bodiless-search/src/LunrSearch.ts:
```diff
@@ -99,6 +99,8 @@ class LunrSearch implements SearchEngineInterface {
      */
     const builder = new Builder();
 
+    const filter = (str: string) => str.replace(/:|\$|#|@|!|\^|\*|\+|-|~|%/gi, ' ');
```
tentatively, add a comment to the function

dewen replied:
fixed

beliayeu commented on packages/bodiless-search/src/SearchClient.ts:
```diff
@@ -41,7 +41,12 @@ class SearchClient implements SearchClientInterface {
+    return this.searchEngine.search(filtered);
+  };
+
+  private filter = (qs: string) => qs.replace(/:|\$|#|@|!|\^|\*|\+|-|~|%/g, ' ');
```
@dewen it looks logic duplication with LunrSearch.ts

dewen replied:
This client side filter is used to remove characters with special meaning to search engine. It is not necessary to be the same as filter inside LunrSearch, which use to tokenize the search content.

We have a ticket gap ticket to follow up on 2nd case.

So I prefer not make this refactoring, rather commented the usage of filter.

beliayeu replied:
@dewen, thanks

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -164,17 +165,28 @@ const SearchBoxBase: FC<SearchProps> = ({ components, ...props }) => {
+   */
+  useEffect(() => {
+    if (queryString !== searchResultContext.searchTerm) {
+      setQueryString(searchResultContext.searchTerm);
```
concerning that we store searchTeam/queryString in 2 places.
any reasons we store it in SearchBoxBase? can we store it in the context and have SearchBoxBase stateless?

dewen replied:
updated

beliayeu commented on packages/bodiless-search/src/components/SearchContextProvider.tsx:
```diff
@@ -44,20 +44,21 @@ export const SearchResultProvider: FC = ({ children }) => {
-      const { q } = querystring.parseUrl(window.location.search).query;
+      const q = querystring.parseUrl(window.location.href, {
+        parseFragmentIdentifier: true,
+      }).fragmentIdentifier;
```
won't it impact anchor links?

dewen replied:
For now, search page doesn't have another fragment usage, it might be used on non-search page, I haven't seen the impact.

See logic:
https://github.com/johnsonandjohnson/Bodiless-JS/pull/592/files#diff-8b9956e79af726fc8e84d6993ec5deaf93770c5aade52de5d001c5bc5be336e7R182

beliayeu commented on packages/bodiless-search/src/components/Search.tsx:
```diff
@@ -164,17 +165,28 @@ const SearchBoxBase: FC<SearchProps> = ({ components, ...props }) => {
   const searchHandler = useCallback(async () => {
     searchLocationValidate();
     searchResultContext.setSearchTerm(queryString);
+    window.location.href = `/${searchPagePath}/#${encodeURIComponent(queryString)}`;
```
why are changing in the location in 2 places? do we need the one in searchLocationValidate?

dewen replied:
good catch, moved to context provider.

dewen replied:
Updated, thanks

TatsianaMiraniuk commented:
@dewen 

please see **issue_8:**
Can you please confirm that BODILESS_SEARCH_INDEX_SELECTOR works with other than the default value?

The goal is to have only rc-menu element is indexed
the following configuration is set per your advise:
BODILESS_SEARCH_INDEX_SELECTOR='rc-menu'
BODILESS_SEARCH_INDEX_EXCLUDE_SELECTOR='script,noscript,style'
BODILESS_SEARCH_INDEX_PREVIEW_LENGTH='300'
BODILESS_DOCS_API_ENABLED=1

However, Search shows 0 results if we search by 'Home', or 'Core', or 'Notification', 'Accordion' - all words from rc-menu
but by some reasons it shows the results if we search by 'List', 'Touts'

The followings were tried too, but it did not work
BODILESS_SEARCH_INDEX_SELECTOR='rc-menu-submenu,bg-gray-200'
BODILESS_SEARCH_INDEX_SELECTOR='.rc-menu-submenu,.bg-gray-200'
BODILESS_SEARCH_INDEX_SELECTOR='.rc-menu'
BODILESS_SEARCH_INDEX_SELECTOR='.bg-gray-200'
BODILESS_SEARCH_INDEX_SELECTOR='bg-gray-200'

It would be very helpful if you can provide the examples that does work for the current Test Site.

Thanks!

TatsianaMiraniuk commented:
> @dewen
> 
> please see **issue_8:**
> Can you please confirm that BODILESS_SEARCH_INDEX_SELECTOR works with other than the default value?
> 
> The goal is to have only rc-menu element is indexed
> the following configuration is set per your advise:
> BODILESS_SEARCH_INDEX_SELECTOR='rc-menu'
> BODILESS_SEARCH_INDEX_EXCLUDE_SELECTOR='script,noscript,style'
> BODILESS_SEARCH_INDEX_PREVIEW_LENGTH='300'
> BODILESS_DOCS_API_ENABLED=1
> 
> However, Search shows 0 results if we search by 'Home', or 'Core', or 'Notification', 'Accordion' - all words from rc-menu
> but by some reasons it shows the results if we search by 'List', 'Touts'
> 
> The followings were tried too, but it did not work
> BODILESS_SEARCH_INDEX_SELECTOR='rc-menu-submenu,bg-gray-200'
> BODILESS_SEARCH_INDEX_SELECTOR='.rc-menu-submenu,.bg-gray-200'
> BODILESS_SEARCH_INDEX_SELECTOR='.rc-menu'
> BODILESS_SEARCH_INDEX_SELECTOR='.bg-gray-200'
> BODILESS_SEARCH_INDEX_SELECTOR='bg-gray-200'
> 
> It would be very helpful if you can provide the examples that does work for the current Test Site.
> 
> Thanks!

The correct value is 
BODILESS_SEARCH_INDEX_SELECTOR='.bg-gray-200 *'
or for multiple selectors:
BODILESS_SEARCH_INDEX_SELECTOR='.bg-gray-200 *,.flex-wrap *'

hvanyo commented on packages/bodiless-search/README.md:
```diff
@@ -0,0 +1,167 @@
+
+> Note: for complete search component implementation example, please check [Test-Site/Search Component](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/examples/test-site/src/components/Search/index.tsx).
+
+
```
@dewen  It would be useful to add section on how to modify the strings: resultCountMessage & resultEmptyMessages.

VancheeZze commented:
Re-approving on behalf of @beliayeu.
